Address modifiers

bfrk opened the pull request:
For discussion see https://code.launchpad.net/~bfrk/epics-base/+git/epics-base/+merge/381628
The branch has been rebased from the above MR.

bfrk committed:
add support for array address modifiers on put

This is independent from server-side filters. It only works with the bracket
notation "PV[s:i:e]", and not with the JSON filter syntax. However, we
re-use the function wrapArrayIndices from the array filters, moving it to
the ioc/db layer.

bfrk committed:
add tests for address modifiers in output links

bfrk committed:
factor out dbHandleModifier from dbPutModifier

This is the first step in changing address modifiers to be polymorphic.

bfrk committed:
address modifier API is now polymorphic

There is still only one implementation, the array range modifier, but
alternative ones could be added now. The concrete behavior of the array
range modifier is unchanged and merely hidden in a module. However parsing
is still part of the dbChannel module. Disentangling this part from
dbChannel will require a bit more restructuring.

Since the generic dbAddrModifier now has a small constant size (two
pointers), it makes more sense to embed it fully inside dbChannel.

bfrk committed:
fix API docs in arrayRangeModifier.h

bfrk committed:
move array range parsing to arrayRangeModifier module

bfrk committed:
fix (justified) warnings in dbUnitTest.c

bfrk commented:
It turned out I had rebased this branch from an out-dated commit. It also turned out that this PR does in fact _not_ depend on the scalar-get-optimization branch, so I have removed this dependency.

bfrk commented:
I don't understand the build failures.

ralphlange commented:
Those?

```
  make[5]: *** [../../../../../configure/RULES_BUILD:259: arrayRangeModifier.obj] Error 2
  make[5]: Entering directory 'D:/a/epics-base/epics-base/modules/database/src/ioc/O.windows-x64'
  cl          -DUSE_TYPED_RSET -DUSE_TYPED_DSET     -nologo -FC -D__STDC__=0 -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -DBUILDING_dbCore_API -DBUILDING_dbCore_API   -Ox -GL -Oy-   -W3        -MD -DEPICS_BUILD_DLL -DEPICS_CALL_DLL   -I. -I../O.Common -I. -I. -I.. -I../as -I../bpt -I../db -I../dbStatic -I../dbtemplate -I../misc -I../registry -I../rsrv -I../../../../../include/compiler/msvc -I../../../../../include/os/WIN32 -I../../../../../include         -c ../db/arrayRangeModifier.c
  arrayRangeModifier.c
  D:\a\epics-base\epics-base\modules\database\src\ioc\db\arrayRangeModifier.c(23): warning C4273: 'wrapArrayIndices': inconsistent dll linkage
  make[4]: *** [../../../../configure/RULES_ARCHS:58: install.windows-x64] Error 2
  D:\a\epics-base\epics-base\modules\database\src\ioc\db\arrayRangeModifier.h(23): note: see previous definition of 'wrapArrayIndices'
  make[3]: *** [../../../configure/RULES_DIRS:85: ioc.install] Error 2
  D:\a\epics-base\epics-base\modules\database\src\ioc\db\arrayRangeModifier.c(58): error C2036: 'const void *': unknown size
  D:\a\epics-base\epics-base\modules\database\src\ioc\db\arrayRangeModifier.c(59): error C2036: 'void *': unknown size
  make[2]: *** [../../configure/RULES_DIRS:85: src.install] Error 2
  make[1]: *** [../configure/RULES_DIRS:85: database.install] Error 2
```

bfrk commented:
Ah, thanks, I saw the bottom part of the failure details and concluded wrongly that it was concerned with the tests. So, yes, those errors. Probably something wrong about the `epicsShareFunc`.

bfrk commented:
I am always confused about `epicsShareFunc` and where and where not to add it. Should I follow the example of e.g. dbEvent.h where it says:
```
#ifdef epicsExportSharedSymbols
#   undef epicsExportSharedSymbols
#   define INCLdbEventhExporting
#endif

#include "epicsThread.h"

#ifdef INCLdbEventhExporting
#   define epicsExportSharedSymbols
#endif

#include "shareLib.h"
```
?

bfrk committed:
dll-export only wrapArrayIndices, not the other functions in arrayRangeModifier.h

bfrk committed:
yet another attempt at fixing the windows build errors

bfrk committed:
fix: remove a gnu-ism in handleArrayRangeModifier

C doesn't allow pointer arithmetic with void* and for good reasons.

bfrk commented:
Okay, looks better. I guess it is ready for review then.

mdavidsaver commented:
> ... Should I follow the example of ...

This is no longer necessary for `epicsThread.h` as the libCom headers are decorated with `LIBCOM_API` instead of `epicsShare*`.

mdavidsaver commented on modules/database/test/std/rec/addrModifierTest.c:
```diff
@@ -0,0 +1,247 @@
+    testdbGetFieldEqual(fieldname, DBF_LONG, NO_ALARM);
+}
+
+#if 0
```
Is this intended as a TODO?

bfrk replied:
I don't think so. It's been a while since I wrote that so not 100% sure. My current best guess is left-over from copy-pasting from other tests. Will delete.

mdavidsaver commented on modules/database/test/std/rec/addrModifierTest.c:
```diff
@@ -0,0 +1,247 @@
+    testdbGetArrFieldEqual("tgt.VAL", DBF_DOUBLE, 10, 8, buf);
+}
+
+#if 0
```
... also

bfrk replied:
> > ... Should I follow the example of ...
> 
> This is no longer necessary for `epicsThread.h` as the libCom headers are decorated with `LIBCOM_API` instead of `epicsShare*`.

I meant the general procedure, not the particular header file. In abf9e68a878 I tried to follow this example. This made some of the windows build problems go away, but I readily admit I have no idea what I am doing here Any advice is welcome. Is there a general guide-line I can study?

mdavidsaver commented on modules/database/src/ioc/db/Makefile:
```diff
@@ -11,10 +11,12 @@
 
 SRC_DIRS += $(IOCDIR)/db
 
+INC += arrayRangeModifier.h
```
Does `arrayRangeModifier.h` need to be public API?

bfrk replied:
`wrapArrayIndices` is used in the array filter code, otherwise: no. I could give this function its own C and header file. Or perhaps you have a better idea?

mdavidsaver replied:
If this header doesn't need to be public, then simply remove the `INC += `.

bfrk committed:
remove some copy-paste left-overs from addrModifierTest.c

mdavidsaver commented:
The ugly undef/def pattern is no longer needed around any of the libCom headers.  At some point we'll convert dbCore and dbRecStd to use the same `*_API` convention, and then these include ordering hacks will be a thing of the past.

wrt. removing `epicsShareFunc` from the functions in `arrayRangeModifier.h` w/o breaking anything.  This confirms that the header should not be installed.  Functions in public headers must be decorated with `epicsShareFunc` or some `*_API` macro.

mdavidsaver commented:
fyi. I do my development builds with:

```
$ cat configure/CONFIG_SITE.local
USR_CFLAGS_Linux += -fvisibility=hidden
USR_CXXFLAGS_Linux += -fvisibility=hidden -fvisibility-inlines-hidden
```

Which can catch some of the situations which lead to DLL linking errors on Windows, and give an idea where `epicsShareFunc` or `*_API` decorations are needed.

bfrk commented:
Am 01.04.21 um 18:57 schrieb mdavidsaver:
> The ugly undef/def pattern is no longer needed around any of the libCom headers.  At some point we'll convert dbCore and dbRecStd to use the same `*_API` convention, and then these include ordering hacks will be a thing of the past.

Good, thanks.

> wrt. removing `epicsShareFunc` from the functions in `arrayRangeModifier.h` w/o breaking anything.  This confirms that the header should not be installed.  Functions in public headers must be decorated with `epicsShareFunc` or some `*_API` macro.

Any suggestion what to do about `wrapArrayIndices`?

mdavidsaver commented:
... I had forgotten that `dbCoreAPI.h` and `DBCORE_API` already exist, and can be used in code under `modules/database/src/ioc/`.  Coupled with `-fvisibility=hidden`, this should catch most if not all situations which would lead to dllimport/export errors with MSVC.

bfrk committed:
remove arrayRangeModifier.h from public API

The only function we export is the utility function wrapArrayIndices, which
now has its own header file. Also use dbCoreAPI.h and DBCORE_API instead of
shareLib.h and epicsShareFunc.

mdavidsaver commented:
> Any suggestion what to do about `wrapArrayIndices`?

Include `dbCoreAPI.h`, and decorate with `DBCORE_API` to export it from dbCore.  This will allow it to be used from dbRecStd.  You may need to add `USR_CPPFLAGS=-I$(TOP)/modules/database/src/ioc/db` in `modules/database/src/std/Makefile` to reach the now private `arrayRangeModifier.h`.

mdavidsaver commented:
> At some point we'll convert dbCore ...

Actually, there is no reason to wait.  cf. #150

bfrk commented:
Am 01.04.21 um 19:55 schrieb mdavidsaver:
>> Any suggestion what to do about `wrapArrayIndices`?
> 
> Include `dbCoreAPI.h`, and decorate with `DBCORE_API` to export it from dbCore.  This will allow it to be used from dbRecStd.  You may need to add `USR_CPPFLAGS=-I$(TOP)/modules/database/src/ioc/db` in `modules/database/src/std/Makefile` to reach the now private `arrayRangeModifier.h`.

Okay, I am now using DBCORE_API. I opted to give wrapArrayIndices its
own header file and install only that. Declaring functions in a header
that cannot be called feels wrong.

mdavidsaver commented:
So you intend `wrapArrayIndices()` to become a public API?

bfrk commented:
Am 01.04.21 um 20:35 schrieb mdavidsaver:
> So you intend `wrapArrayIndices()` to become a public API?

I was operating under the assumption that this is the only way this
function can be shared between the filters and dbCore. Was that wrong?

mdavidsaver commented:
Private headers should not be installed, but rather should be referenced though `USR_CPPFLAGS+=-I...`.

eg.
https://github.com/epics-base/epics-base/blob/c829ca39d873d59947d851a7b22d74d14748e336/modules/database/test/ioc/db/Makefile#L14-L15

bfrk commented:
> Private headers should not be installed, but rather should be referenced though `USR_CPPFLAGS=+-I...`.

Fine with me, I can do that. But I still need to include dbCoreAPI.h and
declare the function DBCORE_API, else I get problems on Windows, right?

mdavidsaver commented:
Yes.

bfrk committed:
treat wrapArrayIndices.h as internal header

bfrk commented:
Okay, done. If you ask me, declaring something as DBCORE_API without it
actually being part of the (public) API feels like lying to me. Perhaps
epicsShareFunc wasn't that bad a name after all.

bfrk commented:
> USR_CFLAGS_Linux += -fvisibility=hidden
> USR_CXXFLAGS_Linux += -fvisibility=hidden -fvisibility-inlines-hidden

BTW, many thanks for that hint, very useful. Is there any reason not to add that to the Linux/gcc config by default?

anjohnson commented:
> Is there any reason not to add that to the Linux/gcc config by default?

Getting the decorations right only matters for applications that need to be supported on Windows. Putting those flags in Base might cause unnecessary problems when building local EPICS applications, and I'd rather we didn't expose our users to the kinds of errors they can cause unnecessarily.

As a developer if you want these and other flags to always apply to your personal builds you can put them in your home/configure directory, I have this file in mine:

```makefile
$ cat ~/configure/CONFIG_USER.Common.linux-x86_64
# target-specific config for linux-x64 targets

# This is needed on RHEL
COMMANDLINE_LIBRARY = READLINE_NCURSES

OP_SYS_CFLAGS += -Wpointer-arith

ifdef BASE_3_15
  OP_SYS_CFLAGS += -fvisibility=hidden
  OP_SYS_CXXFLAGS += -fvisibility=hidden -fvisibility-inlines-hidden
endif

USE_POSIX_THREAD_PRIORITY_SCHEDULING = YES
```

See also `ENABLE_ASAN=YES` if you have a new enough gcc.

mdavidsaver commented:
> Base might cause unnecessary problems when building local EPICS applications

I can erase the "might" from this.  The site local code I've seen mostly doesn't bother with `epicsShare*`.

> ... I'd rather we didn't expose our users to the kinds of errors they can cause unnecessarily.

DItto.  It's hard enough to motivate sites to upgrade as it is.

bfrk commented:
> ... I'd rather we didn't expose our users to the kinds of errors they can cause unnecessarily.

Understood. Thanks for explaining.

mdavidsaver commented on modules/database/src/ioc/db/dbAccess.c:
```diff
@@ -1332,25 +1345,33 @@ long dbPut(DBADDR *paddr, short dbrType,
         /* paddr->pfield may be modified */
         if (status) goto done;
-        if (no_elements < nRequest)
-            nRequest = no_elements;
```
I guess this bounds limit is now expected to be enforced elsewhere?

bfrk replied:
Yes, it is done now immediately before the call to `dbPutConvertRoutine` (for the other cases it is not needed).

mdavidsaver commented on modules/database/src/ioc/db/dbAddrModifier.h:
```diff
@@ -0,0 +1,48 @@
+ */
+typedef struct {
+    void *pvt;                      /** @brief Private modifier data */
+    dbHandleAddrModifier *handle;   /** @brief Function to handle the modifier */
```
It looks like this function pointer will always be `&handleArrayRangeModifier` at present.  Are alternatives possible?  Maybe a premature abstraction?

bfrk replied:
I tend to agree, but see my other reply regarding the extension point.

mdavidsaver commented on modules/database/src/ioc/db/dbChannel.h:
```diff
@@ -54,6 +55,8 @@ typedef struct chFilter chFilter;
     const char *name;
     dbAddr addr;              /* address structure for record/field */
+    dbAddrModifier addrModifier;
+                              /* optional: which indices are targeted */
```
Could `dbHandleAddrModifier` be embedding directly in `dbChannel`?  Seems like this could avoid the need for a new public `dbPutModifier()` and similar.

bfrk replied:
Again, see my general comment about the extension point.

mdavidsaver reviewed (changes requested):
Looking again, I'm wondering about the design of `dbAddrModifier` and `dbHandleAddrModifier`, which are embedded in `dbChannel`, but also handled separately.  The inclusion of a function pointer in `dbAddrModifier` suggests an extension point.  However, limiting to a single function pointer gives no obvious way to handle chaining (as with Get/Monitor filters).  Adding this, perhaps insufficient, extension point now seems to me likely to complicate any later project to add filter-like chaining for Put.  This indirection also seems to be motivating much of the new public API, which might otherwise be avoided.

mdavidsaver commented on modules/database/test/std/rec/addrModifierTest.c:
```diff
@@ -0,0 +1,223 @@
+}
+
+static double initial[] = {0,1,2,3,4,5,6,7,8,9};
+static double buf[10];
```
Looks like this variable could become a local of `expectRange()`.

bfrk replied:
Keeping the declaration at the top level, right after that of `initial`, was deliberate because they have to have the same number of elements.

mdavidsaver replied:
Have a look at `NELEMENTS(X)`, which is our name for the `sizeof(X)/sizeof((X)[0])` trick.

```c
#include <dbDefs.h>
...
    double buf[NELEMENTS(initial)];
```

bfrk replied:
Nice. Fixed in 72c5cf6a8e3

mdavidsaver reviewed (commented):
The added test coverage looks adequate.

bfrk commented:
I agree with your reservations about the extension point. I added that in response to a suggestion by Andrew, see the discussion right at the start of https://code.launchpad.net/~bfrk/epics-base/+git/epics-base/+merge/381628, in particular my reply to Andrew here: https://code.launchpad.net/~bfrk/epics-base/+git/epics-base/+merge/381628/comments/1002334

mdavidsaver commented:
We'll have to find out what @anjohnson thinks of this idea now that you have made it concrete.

bfrk committed:
addrModifierTest.c: move buf into expectRange, use NELEMENTS macro

bfrk commented:
I am trying to rebase to resolve the conflicts. I first checked out 7.0 and did a git pull. Then I did `git rebase 7.0 address-modifiers`. As expected I am getting conflicts. However, when the rebase stops, prompting me to resolve conflicts, I also see this completely unrelated change:
```
diff --git a/.ci b/.ci
index 246428427..3db08b597 160000
--- a/.ci
+++ b/.ci
@@ -1 +1 @@
-Subproject commit 2464284271581a66cb92b704bd9b7c4039282b99
+Subproject commit 3db08b5977efd7a3ac8eecfa068b5bfdfdf84dd0
```
What should I do with that?

ralphlange commented:
That's an update (or not update) of `ci-scripts`.
(See https://github.com/epics-base/ci-scripts/commits/master)
I would not update `ci-scripts` as part of this PR unless it is required.

bfrk commented:
Ah, got it. I need to do `git pull --recurse-submodules` when updating my 7.0 branch. Now the diff is gone.

bfrk commented:
> We'll have to find out what @anjohnson thinks of this idea now that you have made it concrete.

Yes, that would be helpful.

anjohnson commented:
Core Group 2022-10-26: We like the idea of adding this functionality, but we are concerned about committing to API changes which are not extensible to support chaining, as that would probably complicate doing that significantly if/when we do come up with such a design.

mdavidsaver commented:
It occurs to me that since `struct dbChannel` is always allocated in Base, we could add hidden the additional members in super struct.  As is currently done with `struct dbCommonPvt`.

I think this would avoid public API changes, so the only user visible change would be the `.[]` syntax for channel names.  Which seems reasonable since we already support this as a special case for get/monitor.
